pkg/gardenlet/controller/shoot/shoot_control_prepare_migration.go

```diff
@@ -0,0 +1,386 @@
+	}
+
+	o, err := operation.New(shoot, shoot.Status.SeedName, c.config, logger, c.k8sGardenClient, c.k8sGardenCoreInformers.Core().V1beta1(), c.identity, c.secrets, c.imageVector)
+	if err != nil {
+		return reconcile.Result{}, utilerrors.WithSuppressed(err, c.updateShootStatusError(shoot, fmt.Sprintf("Could not initialize a new operation for preparation of Shoot control plane migration: %s", err.Error())))
+	}
+
+	c.recorder.Event(shoot, corev1.EventTypeNormal, gardencorev1beta1.EventPrepareMigration, "Prepare Shoot cluster for migration")
+	if err := c.updateShootStatusPrepareMigrationStart(o); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := c.runPrepareShootControlPlaneMigration(o); err != nil {
+		c.recorder.Event(shoot, corev1.EventTypeWarning, gardencorev1beta1.EventMigrationPreparationFailed, err.Description)
+		return reconcile.Result{}, utilerrors.WithSuppressed(errors.New(err.Description), c.updateShootStatusPrepareMigrationError(o, err.Description, err.LastErrors...))
+	}
+
+	return c.finalizeShootPrepareForMigration(shoot, o)
+}
+
+func (c *Controller) runPrepareShootControlPlaneMigration(o *operation.Operation) *gardencorev1beta1helper.WrappedLastErrors {
+	var (
+		namespace       = &corev1.Namespace{}
+		botanist        *botanistpkg.Botanist
+		err             error
+		dnsEnabled      = !o.Shoot.DisableDNS
+		shootHibernated = isShootHibernated(o.Shoot.Info.Spec.Hibernation)
+		tasksWithErrors []string
+	)
+
+	for _, lastError := range o.Shoot.Info.Status.LastErrors {
+		if lastError.TaskID != nil {
+			tasksWithErrors = append(tasksWithErrors, *lastError.TaskID)
+		}
+	}
+
+	errorContext := utilerrors.NewErrorContext("Shoot's control plane preparation for migration", tasksWithErrors)
+
+	err = utilerrors.HandleErrors(errorContext,
+		func(errorID string) error {
+			o.CleanShootTaskError(context.TODO(), errorID)
+			return nil
+		},
+		nil,
+		utilerrors.ToExecute("Create botanist", func() error {
+			return retryutils.UntilTimeout(context.TODO(), 10*time.Second, 10*time.Minute, func(context.Context) (done bool, err error) {
+				botanist, err = botanistpkg.New(o)
+				if err != nil {
+					return retryutils.MinorError(err)
+				}
+				return retryutils.Ok()
+			})
+		}),
+		utilerrors.ToExecute("Retrieve the Shoot namespace in the Seed cluster", func() error {
+			err := botanist.K8sSeedClient.Client().Get(context.TODO(), client.ObjectKey{Name: o.Shoot.SeedNamespace}, namespace)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					o.Logger.Infof("Did not find '%s' namespace in the Seed cluster - nothing to be done", o.Shoot.SeedNamespace)
+					return utilerrors.Cancel()
+				}
+			}
+			return err
+		}),
+	)
+
+	if err != nil {
+		if utilerrors.WasCanceled(err) {
+			return nil
+		}
+		return gardencorev1beta1helper.NewWrappedLastErrors(gardencorev1beta1helper.FormatLastErrDescription(err), err)
+	}
+
+	var (
+		nonTerminatingNamespace = namespace.Status.Phase != corev1.NamespaceTerminating
```

Comment:
Why do we need this?

Reply:
If namespace is set to deletion we cannot redeploy the etcd.

Reply:
Sorry for the late response, but how can this happen? The shoot namespace in the old seed can only be in terminating state if a shoot deletion was triggered (but then we can't trigger a migration), or if another migration was triggered previously (but then we can't trigger a new one as this seed is no longer the source).

Reply:
What about, if it fails during the namespace deletion, if it times out while waiting?

Reply:
OK, that's a good point. Thank you! :)